torch/csrc/distributed/c10d/intra_node_comm.cpp

```diff
@@ -0,0 +1,628 @@
+
+    Shared(size_t worldSize) : twoPhaseExchange(worldSize) {
+      deviceIds.fill(-1);
+    }
+  };
+
+  // Initialize shared memory for rendezvous
+  SharedMemoryPtr<Shared> peerInfo(rdzvId, rank, worldSize, worldSize);
+
+  // Gether peer device Ids
+  std::vector<int> deviceIds;
+  peerInfo->twoPhaseExchange.run(
+      [&]() { peerInfo->deviceIds[rank] = at::cuda::current_device(); },
+      [&]() {
+        for (const auto& deviceId : peerInfo->deviceIds) {
+          if (deviceId != -1) {
+            deviceIds.push_back(deviceId);
+          }
+        }
+      });
+  TORCH_CHECK(
+      deviceIds.size() == worldSize,
+      "More than one rank is using the same device. "
+      "Make sure every rank is assigned with a unique device. ");
+
+  // Query nvlink connection
+  auto nvlMesh = getNvlMesh(worldSize);
+
+  // Detect topology
+  Topology topology = detectTopology(nvlMesh, deviceIds);
+  if (topology == Topology::UNKNOWN) {
+    return nullptr;
+  }
+
+  // Initialize p2p state
+  auto p2pState = initP2pState(topology, nvlMesh, rank);
+
+  // Allocate buffer
+  void* buffer = nullptr;
+  C10_CUDA_CHECK(cudaMalloc(&buffer, kMaxIntraNodeSize * 2));
+
+  // Make p2p state and buffer available for IPC
+  cudaIpcMemHandle_t p2pStateHandle, bufferHandle;
+  AT_CUDA_CHECK(cudaIpcGetMemHandle(&p2pStateHandle, p2pState));
+  AT_CUDA_CHECK(cudaIpcGetMemHandle(&bufferHandle, buffer));
+
+  // Exchange IPC handles for p2p state and buffer
+  std::array<cudaIpcMemHandle_t, kMaxDevices> p2pStateHandles, bufferHandles;
+  peerInfo->twoPhaseExchange.run(
+      [&]() {
+        peerInfo->p2pStateHandles[rank] = p2pStateHandle;
+        peerInfo->bufferHandles[rank] = bufferHandle;
+      },
+      [&]() {
+        p2pStateHandles = peerInfo->p2pStateHandles;
+        bufferHandles = peerInfo->bufferHandles;
+      });
+
+  // Map peer p2p states and buffers to the process address space
+  std::array<void*, kMaxDevices> p2pStates, buffers;
+  for (size_t r = 0; r < worldSize; ++r) {
+    if (r == rank) {
+      p2pStates[r] = p2pState;
+      buffers[r] = buffer;
+    } else {
+      AT_CUDA_CHECK(cudaIpcOpenMemHandle(
+          &p2pStates[r], p2pStateHandles[r], cudaIpcMemLazyEnablePeerAccess));
+      AT_CUDA_CHECK(cudaIpcOpenMemHandle(
+          &buffers[r], bufferHandles[r], cudaIpcMemLazyEnablePeerAccess));
+    }
+  }
+  return c10::make_intrusive<IntraNodeComm>(
+      topology, p2pStates, buffers, rank, worldSize);
+#else
+  return nullptr;
+#endif
+}
+
+c10::intrusive_ptr<IntraNodeComm> IntraNodeComm::rendezvousViaStore(
+    c10::intrusive_ptr<c10d::Store> store,
+    const std::string& prefix,
+    size_t rank,
+    size_t worldSize) {
+#ifndef USE_ROCM
+  if (!isIntraNodeCommSupported() ||
+      !getCvarBool(ENABLE_INTRA_NODE_COMM, false) || worldSize < 2 ||
+      worldSize > kMaxDevices) {
+    return nullptr;
+  }
+  std::array<char, HOST_NAME_MAX + 1> buf = {};
+  gethostname(buf.data(), buf.size());
```

Comment:
Personally I wouldn't trust hostnames enough to use them to detect if two ranks are on the same node. I would recommend using the content of /proc/sys/kernel/random/boot_id. See https://man7.org/linux/man-pages/man4/random.4.html